src/Symfony/Component/DomCrawler/DomCrawler.php

```diff
@@ -0,0 +1,350 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DomCrawler;
+
+use Masterminds\HTML5;
+
+/**
+ * DomCrawler eases navigation of a list of \DOMNode or \DOM\Node objects.
+ *
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @implements \IteratorAggregate<int, \DOMNode>|\IteratorAggregate<int, \DOM\Node>
+ */
+class DomCrawler extends Crawler
+{
+    /**
+     * Whether to parse HTML5 content or consider everything as XML.
+     */
+    public const CRAWLER_ENABLE_HTML5_PARSING = 0b1;
+
+    /**
+     * Use an external parser to parse the HTML content, or libxml for XML content.
+     */
+    public const CRAWLER_USE_EXTERNAL_PARSER = 0b10;
+
+    /**
+     * Available since PHP 8.4 to use the native DOM extension parser.
+     * The native parser provides better performance but strictly follows
+     * the XML specification, which may lead to different results compared
+     * to the external parser or libxml.
+     */
+    public const CRAWLER_USE_NATIVE_PARSER = 0b100;
+
+    /**
+     * @param \DOMNodeList|\DOM\NodeList|\DOMNode|\DOM\Node|\DOMNode[]|\DOM\Node[]|string|null $node    A Node to use as the base for the crawling
+     * @param int                                                                              $options A bitmask of CRAWLER_* constants
+     */
+    public function __construct(
+        \DOMNodeList|\DOM\NodeList|\DOMNode|\DOM\Node|array|string|null $node = null,
+        ?string $uri = null,
+        ?string $baseHref = null,
+        protected int $options = self::CRAWLER_ENABLE_HTML5_PARSING | self::CRAWLER_USE_EXTERNAL_PARSER,
+    ) {
+        if ($this->options & self::CRAWLER_USE_NATIVE_PARSER && \PHP_VERSION_ID < 80400) {
+            throw new \InvalidArgumentException('Native parser requires PHP 8.4 or higher.');
+        }
+
+        if ($this->options & self::CRAWLER_USE_EXTERNAL_PARSER && $this->options & self::CRAWLER_USE_NATIVE_PARSER) {
+            throw new \InvalidArgumentException('You cannot use both external and native parsers at the same time.');
+        }
+
+        if ($this->options & self::CRAWLER_ENABLE_HTML5_PARSING && !($this->options & self::CRAWLER_USE_EXTERNAL_PARSER) && !($this->options & self::CRAWLER_USE_NATIVE_PARSER)) {
```

Comment:
Is the parsing behavior for HTML different between the native parser of PHP 8.4 and masterminds/html5 ?

Reply:
Yes, the new native parser follows specifications strictly.

This is why there is a "separate test suite" for it, because many dataset of the current tests are not "valid strict HTML": unclosed tags, unneeded slashes to close some tags... many cases where the new parser throws at parsing. This is why I opted for an opt-in update. Devs will have to fully test their app with this new parser before using it extensively.

Reply:
The masterminds/html5 parser is also supposed to follow the spec.
And currently, the opt-in we use to switch between the libxml parser and the HTML5 parser is the HTML5 doctype.

Reply:
Here's a snippet comparing ext-dom and masterminds/html5:

```php
require __DIR__.'/vendor/autoload.php';

$html = '<!DOCTYPE html><html><base href="https://base.com/"><form method="post" action="link/"><button type="submit" name="submit" /></form></html>';

$parser = new \Masterminds\HTML5();
$oldDom = $parser->loadHTML($html);
dump($oldDom->childNodes->item(1)->childNodes->item(1)->tagName);

$newDom = \DOM\HTMLDocument::createFromString($html);
dump($newDom->childNodes->item(1)->childNodes->item(1)->tagName);
```

The output is:

```
"form"
PHP Warning:  DOM\HTMLDocument::createFromString(): tree error unexpected-element-in-open-elements-stack in Entity, line: 1, column: 128-131 in /Users/alex/PhpstormProjects/symfony/test.php on line 11

Warning: DOM\HTMLDocument::createFromString(): tree error unexpected-element-in-open-elements-stack in Entity, line: 1, column: 128-131 in /Users/alex/PhpstormProjects/symfony/test.php on line 11
"BODY"
```

Note that ext-dom also throws a warning because of the `button` tag being closed with `/>` instead of `</button>`.